Checkout: add window.location redirect fallback after payment complete

sirbrillig opened the pull request:
#### Changes proposed in this Pull Request

The checkout thank-you URL can be many different things, and occasionally it's not a wordpress.com URL at all (eg: for a Jetpack site). Currently, `useCreatePaymentCompleteCallback` uses `page.redirect(url)` to redirect after checkout completes, but it appears that `page.redirect()` uses `History.replaceState()` which cannot be used for cross-origin changes. In fact, we don't really want to use `History.replaceState()` anyway because there's really no need to remove checkout from the browser history.

This PR changes `useCreatePaymentCompleteCallback` to use `page()` instead, which still has a similar issue because it uses `History.pushState()`, so it adds a fallback to use `window.location.href = url` instead if `page()` fails. The reason we need to use `page()` in the first place is that there are several notices that may be displayed when checkout completes, and `page()` redirects to calypso without reloading the page, allowing those notices to persist to the next page. Of course, it's fine if they don't persist if the redirect URL is not in calypso, but we want to give them a chance.

Note that the post-checkout notices were actually broken because of mistyped data from the endpoint. This PR also fixes them. For example, if you renew a purchase:

<img width="744" alt="Screen Shot 2021-01-27 at 5 18 52 PM" src="https://user-images.githubusercontent.com/2036909/106062518-e19dbe00-60c4-11eb-85c3-c34221bd4b58.png">


#### Testing instructions

- Complete checkout with any flow and verify that you are successfully redirected to a new page afterward.

sirbrillig committed:
Checkout: redirect using window.location

elliottprogrammer reviewed (approved):
Tested this with the `redirectTo` prop and it works now!  Thanks very much for your help with this @sirbrillig 👍   LGTM!

sirbrillig committed:
Replace other use of page.redirect in callback

sirbrillig committed:
Update persistent option to displayOnNextPage

The API was updated but this option was not.

sirbrillig commented:
Ah, the only problem with this is that it prevents notices from being displayed when the next page loads. Looking into it...

sirbrillig committed:
Use page for SPA redirects with window.location fallback

Without using `page`, we lose the ability to display messages on the
next page, but this way there's a fallback if `page` fails.

Also note that `page()` instead of `page.redirect()` does not use
`replaceState` although it does use `pushState` which is vulnerable to
similar issues.

This restores the required use of `window.location` for new accounts.

sirbrillig committed:
Fix purchases in TransactionResponse type

sirbrillig committed:
Fix flattening purchases result when displaying checkout success

sirbrillig commented:
Ok, I think it should work for all cases now 🤞 . I had to restore using `page` (although not `page.redirect`) because actually reloading the page breaks persistent notices, but I added `window.location` as a fallback.

elliottprogrammer reviewed (approved):
I tested checkout with `redirectTo` to the site's `wp-admin/post.php?post=1` and also tested a standard checkout with a Jetpack product where it redirected to the fallback thankYou page. It all seems to be working good. LGTM! 👍
